sphinx-js incompatible with Sphinx 4.1.0+

Sphinx-js cannot simply update to use the latest version of sphinx-doc. This is preventing sphinx-js from supporting python 3.10

When we use that commit or later versions of Sphinx the sphinx-js test suite incorrectly prints argument, error, and return types as `"type()"` instead of the expected `*type*`. It looks something like this:

```
tests/test_build_js/test_build_js.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test_build_js.Tests testMethod=test_autofunction_callback>
filename = 'autofunction_callback'
contents = 'requestCallback(responseCode)\n\n   Some global callback\n\n   Arguments:\n      * **responseCode** (*number*) --\n'

    def _file_contents_eq(self, filename, contents):
>       assert self._file_contents(filename) == contents
E       AssertionError: assert ('requestCallback(responseCode)\n'\n '\n'\n '   Some global callback\n'\n '\n'\n '   Arguments:\n'\n '      * **responseCode** ("number()") --\n') == ('requestCallback(responseCode)\n'\n '\n'\n '   Some global callback\n'\n '\n'\n '   Arguments:\n'\n '      * **responseCode** (*number*) --\n')
E           requestCallback(responseCode)
E           
E              Some global callback
E           
E              Arguments:
E         -       * **responseCode** (*number*) --
E         ?                           ^      ^
E         +       * **responseCode** ("number()") --
E         ?                           ^      ^ ++
```

After [some investigation](https://github.com/mozilla/sphinx-js/pull/191), I've narrowed the incompatibility to Sphinx-doc 4.1.0 and newer, specifically https://github.com/sphinx-doc/sphinx/commit/4534d2d1a5755c8cbc9ef4327eab7e34a85a7de8. This was shipped as part of [sphinx-doc PR 9155](https://github.com/sphinx-doc/sphinx/pull/9155), which resolved [Sphinx-doc issue 8945](https://github.com/sphinx-doc/sphinx/issues/8945#issuecomment-812981784).

The commit contains a change to the [docfields.py:make_xref function](https://github.com/sphinx-doc/sphinx/commit/4534d2d1a5755c8cbc9ef4327eab7e34a85a7de8#diff-f18bf3af0871669d46455026b741254522ecd7e8a729be708603249e872d37abR66-R94). If I revert just this portion of the change things work as expected.

In the previous code the role is present but ignored. The prior code path is preserved there but there are no conditions under which it can be reached. The inliner is always passed along with domain and role ('js:function'). The test failures we see are all `xref` objects that resolve through the new code as functions, which is why they get inappropriately rendered with parenthesis.

It's not obvious to me what should or even can be done about this from the sphinx-js side of the code.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -16,15 +16,15 @@ jobs:
 
     strategy:
       matrix:
-        python-version: ['3.7', '3.8', '3.9']
+        python-version: ['3.7', '3.8', '3.9', '3.10', '3.11']
         experimental: [false]
 
     name: Python ${{ matrix.python-version}}
     steps:
-      - uses: actions/checkout@v2.3.5
+      - uses: actions/checkout@v3.1.0
 
       - name: Set up Python
-        uses: actions/setup-python@v2.2.2
+        uses: actions/setup-python@v4.3.0
         with:
           python-version: ${{ matrix.python-version }}
 

--- a/requirements_dev.txt
+++ b/requirements_dev.txt
@@ -1,4 +1,4 @@
-pytest==6.2.3
+pytest==7.2.0
 recommonmark==0.7.1
 # Sphinx's plain-text renderer changes behavior slightly
 # with regard to how it emits class names and em dashes from
@@ -6,4 +6,4 @@ recommonmark==0.7.1
 Sphinx==3.5.4
 markupsafe==2.0.1
 Jinja2>2.0,<3.0
-
+tox==4.0.3

--- a/requirements_dev.txt
+++ b/requirements_dev.txt
@@ -1,9 +1,5 @@
+-e .
+
 pytest==7.2.0
 recommonmark==0.7.1
-# Sphinx's plain-text renderer changes behavior slightly
-# with regard to how it emits class names and em dashes from
-# time to time:
-Sphinx==3.5.4
-markupsafe==2.0.1
-Jinja2>2.0,<3.0
 tox==4.0.3

--- a/setup.py
+++ b/setup.py
@@ -14,7 +14,11 @@ setup(
     packages=find_packages(exclude=['ez_setup']),
     url='https://github.com/mozilla/sphinx-js',
     include_package_data=True,
-    install_requires=['Jinja2>2.0,<3.0', 'markupsafe==2.0.1', 'parsimonious>=0.9.0,<0.10.0', 'Sphinx>=3.0.0'],
+    install_requires=[
+        'Jinja2>2.0',
+        'parsimonious>=0.10.0,<0.11.0',
+        'Sphinx>=4.1.0,<6.0.0'
+    ],
     python_requires='>=3.7',
     classifiers=[
         'Intended Audience :: Developers',
@@ -24,6 +28,6 @@ setup(
         'Programming Language :: Python :: 3',
         'Topic :: Documentation :: Sphinx',
         'Topic :: Software Development :: Documentation'
-        ],
+    ],
     keywords=['sphinx', 'documentation', 'docs', 'javascript', 'js', 'jsdoc', 'restructured', 'typescript', 'typedoc'],
 )

--- a/setup.py
+++ b/setup.py
@@ -18,6 +18,9 @@ setup(
         'Jinja2>2.0',
         'parsimonious>=0.10.0,<0.11.0',
         'Sphinx>=4.1.0,<6.0.0'
+        # Pin markupsafe because of
+        # https://github.com/pallets/jinja/issues/1585
+        'markupsafe==2.0.1',
     ],
     python_requires='>=3.7',
     classifiers=[

--- a/sphinx_js/__init__.py
+++ b/sphinx_js/__init__.py
@@ -1,3 +1,4 @@
+from functools import lru_cache
 from os.path import join, normpath
 
 from sphinx.errors import SphinxError
@@ -8,6 +9,68 @@ from .directives import (auto_class_directive_bound_to_app,
                          JSStaticFunction)
 from .jsdoc import Analyzer as JsAnalyzer
 from .typedoc import Analyzer as TsAnalyzer
+
+
+# Cache this to guarantee it only runs once.
+@lru_cache(maxsize=None)
+def fix_js_make_xref():
+    """Monkeypatch to fix sphinx.domains.javascript TypedField and GroupedField
+
+    Fixes https://github.com/sphinx-doc/sphinx/issues/11021
+
+    """
+    from docutils import nodes
+
+    from sphinx.domains import javascript
+    from sphinx.locale import _
+
+    class JSXrefMixin:
+        def make_xref(
+            self,
+            rolename,
+            domain,
+            target,
+            innernode=nodes.emphasis,
+            contnode=None,
+            env=None,
+            inliner=None,
+            location=None,
+        ):
+            # Set inliner to None just like the PythonXrefMixin does so the
+            # xref doesn't get rendered as a function.
+            return super().make_xref(
+                rolename, domain, target,
+                innernode, contnode,
+                env, inliner=None, location=None)
+
+    class JSTypedField(JSXrefMixin, javascript.TypedField):
+        pass
+
+    class JSGroupedField(JSXrefMixin, javascript.GroupedField):
+        pass
+
+    # Replace javascript module
+    javascript.TypedField = JSTypedField
+    javascript.GroupedField = JSGroupedField
+
+    # Fix the one place TypedField and GroupedField are used in the javascript
+    # module
+    javascript.JSCallable.doc_field_types = [
+        JSTypedField(
+            'arguments', label=_('Arguments'),
+            names=('argument', 'arg', 'parameter', 'param'),
+            typerolename='func', typenames=('paramtype', 'type')
+        ),
+        JSGroupedField(
+            'errors', label=_('Throws'), rolename='func',
+            names=('throws', ),
+            can_collapse=True
+        ),
+    ] + javascript.JSCallable.doc_field_types[2:]
+
+
+fix_js_make_xref()
+
 
 def setup(app):
     # I believe this is the best place to run jsdoc. I was tempted to use

--- a/sphinx_js/directives.py
+++ b/sphinx_js/directives.py
@@ -11,6 +11,7 @@ from os.path import join, relpath
 
 from docutils.parsers.rst import Directive
 from docutils.parsers.rst.directives import flag
+from sphinx import addnodes
 from sphinx.domains.javascript import JSCallable
 
 from .renderers import (AutoFunctionRenderer,
@@ -113,4 +114,6 @@ def _members_to_exclude(arg):
 
 class JSStaticFunction(JSCallable):
     """Like a callable but with a different prefix."""
-    display_prefix = 'static '
+    def get_display_prefix(self):
+        return [addnodes.desc_sig_keyword('static', 'static'),
+                addnodes.desc_sig_space()]

--- a/tox.ini
+++ b/tox.ini
@@ -13,9 +13,9 @@ python =
 setenv =
     PATH={toxinidir}/node_modules/.bin{:}{envbindir}{:}{env:PATH}
 deps = -rrequirements_dev.txt
-whitelist_externals =
-  env
-  npm
+allowlist_externals =
+    env
+    npm
 commands_pre = npm install --no-save jsdoc@3.6.3 typedoc@0.15.0
 # Contrary to the tox docs, setenv's changes to $PATH are not visible inside
 # any commands we call. I hack around this with env:

